sdk/src/main/java/io/opentelemetry/sdk/tags/TagMapSdk.java

```diff
@@ -0,0 +1,161 @@
+   * @param parent providing a default set of tags
+   */
+  private TagMapSdk(Map<? extends TagKey, ? extends Tag> tags, TagMap parent) {
+    this.tags = Collections.unmodifiableMap(new HashMap<>(checkNotNull(tags, "tags")));
+    this.parent = parent;
+  }
+
+  @Override
+  public Iterator<Tag> getIterator() {
+    Map<TagKey, Tag> combined = new HashMap<>(tags);
+    if (parent != null) {
+      for (Iterator<Tag> it = parent.getIterator(); it.hasNext(); ) {
+        Tag tag = it.next();
+        if (!combined.containsKey(tag.getKey())) {
+          combined.put(tag.getKey(), tag);
+        }
+      }
+    }
+    // Clean out any null values that may have been added by Builder.remove.
+    for (Iterator<Tag> it = combined.values().iterator(); it.hasNext(); ) {
+      if (it.next() == null) {
+        it.remove();
+      }
+    }
+
+    return Collections.unmodifiableCollection(combined.values()).iterator();
+  }
+
+  @Nullable
+  @Override
+  public TagValue getTagValue(TagKey tagKey) {
+    Tag tag = tags.get(tagKey);
+    if (tag != null) {
+      return tag.getValue();
+    } else {
+      return parent == null ? null : parent.getTagValue(tagKey);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || !(o instanceof TagMapSdk)) {
+      return false;
+    }
+
+    TagMapSdk tagMapSdk = (TagMapSdk) o;
+
+    if (!tags.equals(tagMapSdk.tags)) {
+      return false;
+    }
+    return parent != null ? parent.equals(tagMapSdk.parent) : tagMapSdk.parent == null;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = tags.hashCode();
+    result = 31 * result + (parent != null ? parent.hashCode() : 0);
+    return result;
+  }
+
+  // TODO: Migrate to AutoValue.Builder
+  // @AutoValue.Builder
+  static class Builder implements TagMap.Builder {
+    @Nullable private TagMap parent;
+    private final Map<TagKey, Tag> tags;
+
+    /** Create a new empty TagMap builder. */
+    Builder() {
+      this.tags = new HashMap<>();
+    }
+
+    @Override
+    public TagMap.Builder setParent(TagMap parent) {
+      this.parent = parent;
+      return this;
+    }
+
+    @Override
+    public TagMap.Builder setNoParent() {
+      parent = null;
+      return this;
+    }
+
+    @Override
+    public TagMap.Builder put(TagKey key, TagValue value, TagMetadata tagMetadata) {
+      tags.put(
+          checkNotNull(key, "key"),
+          Tag.create(key, checkNotNull(value, "value"), checkNotNull(tagMetadata, "tagMetadata")));
+      return this;
+    }
+
+    @Override
+    public TagMap.Builder remove(TagKey key) {
+      tags.remove(checkNotNull(key, "key"));
+      if (parent != null && parent.getTagValue(key) != null) {
+        tags.put(key, null);
+      }
+      return this;
+    }
+
+    @Override
+    public TagMapSdk build() {
+      return new TagMapSdk(tags, parent);
+    }
+
+    @Override
+    public Scope buildScoped() {
```

Comment:
Consider to add an issue to remove this to ensure consistency with Span, where we removed the scoped version. May be different arguments here but still.

Reply:
I think it can stay as there are no side effects (i.e. closing `Span`), but still, yeah, lets create an issue for this one.

Reply:
Filled #400 so we can move forward.